File: windows/Tama.Core/OllamaReader.cs
using System.Globalization;
using System.Text;

namespace Tama.Core;

/// <summary>
/// Reads per-model activity for a local Ollama server from its llama.cpp server.log.
/// Read-only and local-only: never opens the Ollama HTTP API, never reads history files.
/// Only the log tail is inspected. Activity is attributed by segmenting the log between
/// runner start/stop events. Mirror of the Swift OllamaReader (spec/log-formats.md, Ollama).
/// </summary>
public sealed class OllamaReader
{
    private const int TailBytes = 256 * 1024;
    private readonly string _logPath;
    private readonly TimeZoneInfo _tz;

    /// <summary>tz parses the [GIN] request timestamps, which carry no zone of their own.</summary>
    public OllamaReader(string logPath, TimeZoneInfo tz)
    {
        _logPath = logPath;
        _tz = tz;
    }

    /// <summary>Default Windows location of the Ollama server log.</summary>
    public static string DefaultLogPath() => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "Ollama", "server.log");

    private sealed class Segment
    {
        public bool Busy;
        public DateTimeOffset? LastActivity;
        public int? ContextWindow;
        public int? ContextTokens;
        public double? TokensPerSecond;
        public double? LastLatencySeconds;
        public int RequestCount;
        public int ChatCount;
        public int EmbedCount;
        public OllamaRequestKind Kind =>
            EmbedCount > ChatCount ? OllamaRequestKind.Embed
            : ChatCount > 0 ? OllamaRequestKind.Chat : OllamaRequestKind.Unknown;
    }

    public OllamaStatus? Read()
    {
        var data = SafeFileReader.Tail(_logPath, TailBytes);
        if (data is not { Length: > 0 }) return null;
        var text = Encoding.UTF8.GetString(data);

        var segments = new Dictionary<string, Segment>();   // keyed by model tag (dedupes reloads)
        var order = new List<string>();                     // first-seen order
        string? active = null;                              // model owning the lines being read
        string? current = null;                             // model whose runner is still loaded

        foreach (var line in text.Split('\n', StringSplitOptions.RemoveEmptyEntries))
        {
            if (StartedModel(line) is { } tag)
            {
                active = tag; current = tag;
                if (!segments.ContainsKey(tag)) { segments[tag] = new Segment(); order.Add(tag); }
                continue;
            }
            if (line.Contains("stopping mlx runner subprocess", StringComparison.Ordinal))
            {
                active = null; current = null;
                continue;
            }
            if (active is null || !segments.TryGetValue(active, out var seg)) continue;
            Apply(line, seg);
        }

        var models = order
            .Select(tag =>
            {
                var s = segments[tag];
                return new OllamaModelActivity(tag, Current: tag == current, Busy: s.Busy,
                    LastActivity: s.LastActivity, ContextWindow: s.ContextWindow,
                    ContextTokens: s.ContextTokens, TokensPerSecond: s.TokensPerSecond,
                    LastLatencySeconds: s.LastLatencySeconds, Kind: s.Kind,
                    RequestCount: s.RequestCount);
            })
            .OrderByDescending(m => m.Current)                                   // loaded model first
            .ThenByDescending(m => m.LastActivity ?? DateTimeOffset.MinValue)
            .ToList();
        return new OllamaStatus(Running: false, LastActivity: ModificationDate(), Models: models);
    }

    private void Apply(string line, Segment seg)
    {
        if (IntAfter("n_ctx_slot = ", line) is { } w) seg.ContextWindow = w;
        if (IntAfter("task.n_tokens = ", line) is { } t) seg.ContextTokens = t;
        if (DoubleAfter("tg = ", line) is { } tps) seg.TokensPerSecond = tps;
        if (line.Contains("all slots are idle", StringComparison.Ordinal)) seg.Busy = false;
        if (line.Contains("processing task", StringComparison.Ordinal)) seg.Busy = true;
        if (line.StartsWith("[GIN]", StringComparison.Ordinal)) ApplyGin(line, seg);
    }

    private void ApplyGin(string line, Segment seg)
    {
        if (GinEndpoint(line) is not { } endpoint) return;
        OllamaRequestKind kind;
        if (endpoint.Contains("/embed", StringComparison.Ordinal)) kind = OllamaRequestKind.Embed;
        else if (endpoint.EndsWith("/api/chat", StringComparison.Ordinal)
              || endpoint.EndsWith("/api/generate", StringComparison.Ordinal)) kind = OllamaRequestKind.Chat;
        else return;   // /api/tags, /show, /ps, /version, /pull, /delete — not inference
        seg.RequestCount++;
        if (kind == OllamaRequestKind.Embed) seg.EmbedCount++; else seg.ChatCount++;
        var fields = line.Split('|');
        if (fields.Length > 2 && ParseDuration(fields[2].Trim()) is { } dur) seg.LastLatencySeconds = dur;
        var stamp = fields[0].Replace("[GIN]", "", StringComparison.Ordinal).Trim();
        if (DateTime.TryParseExact(stamp, "yyyy/MM/dd - HH:mm:ss", CultureInfo.InvariantCulture,
                DateTimeStyles.None, out var local))
            seg.LastActivity = new DateTimeOffset(local, _tz.GetUtcOffset(local));
    }

    /// <summary>model=&lt;tag&gt; on a runner-subprocess start line (ignores pull-manifest noise).</summary>
    private static string? StartedModel(string line)
    {
        if (!line.Contains("runner subprocess", StringComparison.Ordinal)
            || !line.Contains("starting", StringComparison.Ordinal)) return null;
        var i = line.IndexOf("model=", StringComparison.Ordinal);
        if (i < 0) return null;
        var rest = line[(i + "model=".Length)..];
        var end = rest.AsSpan().IndexOfAny(' ', '\t');
        var tag = end < 0 ? rest : rest[..end];
        return tag.Length == 0 ? null : tag;
    }

    private static int? IntAfter(string needle, string line)
    {
        var i = line.IndexOf(needle, StringComparison.Ordinal);
        if (i < 0) return null;
        var rest = line.AsSpan(i + needle.Length);
        var len = 0;
        while (len < rest.Length && char.IsAsciiDigit(rest[len])) len++;
        return len > 0 && int.TryParse(rest[..len], out var v) ? v : null;
    }

    private static double? DoubleAfter(string needle, string line)
    {
        var i = line.IndexOf(needle, StringComparison.Ordinal);
        if (i < 0) return null;
        var rest = line.AsSpan(i + needle.Length).TrimStart(' ');
        var len = 0;
        while (len < rest.Length && (char.IsAsciiDigit(rest[len]) || rest[len] == '.')) len++;
        return len > 0 && double.TryParse(rest[..len], CultureInfo.InvariantCulture, out var v) ? v : null;
    }

    /// <summary>The quoted route on a [GIN] line, e.g. /api/chat.</summary>
    private static string? GinEndpoint(string line)
    {
        var close = line.LastIndexOf('"');
        if (close < 0) return null;
        var open = line.LastIndexOf('"', close - 1);
        if (open < 0) return null;
        return line[(open + 1)..close];
    }

    /// <summary>llama.cpp's Go-duration strings: 5.0s, 200ms, 23.917µs, 41ns.</summary>
    private static double? ParseDuration(string s)
    {
        double? Value(int drop, double scale) =>
            double.TryParse(s[..^drop], CultureInfo.InvariantCulture, out var v) ? v * scale : null;
        if (s.EndsWith("ms", StringComparison.Ordinal)) return Value(2, 1e-3);
        if (s.EndsWith("µs", StringComparison.Ordinal) || s.EndsWith("us", StringComparison.Ordinal)) return Value(2, 1e-6);
        if (s.EndsWith("ns", StringComparison.Ordinal)) return Value(2, 1e-9);
        if (s.EndsWith("s", StringComparison.Ordinal)) return Value(1, 1);
        return null;
    }

    private DateTimeOffset? ModificationDate()
    {
        try
        {
            return File.Exists(_logPath)
                ? new DateTimeOffset(File.GetLastWriteTimeUtc(_logPath), TimeSpan.Zero) : null;
        }
        catch (Exception e) when (e is IOException or UnauthorizedAccessException or ArgumentException or NotSupportedException)
        { return null; }
    }
}

File: Sources/TamaCore/Usage/OllamaReader.swift
import Foundation

/// Supplies Ollama enrichment from the log. A protocol so the monitor can be tested with a stub.
public protocol OllamaReading: Sendable {
    func read() -> OllamaStatus?
}

/// Reads per-model activity for a local Ollama server from its llama.cpp `server.log`.
///
/// Read-only and local-only by construction: it never opens the Ollama HTTP API
/// (`http://localhost:11434/...`) — doing so would break Tama's no-network invariant — and never
/// reads `~/.ollama/history` (plaintext user prompts). Only `server.log` is parsed, via
/// `SafeFileReader`. Presence (`running`) is filled in by the caller from the process table.
///
/// The log carries no per-line model tag, so activity is attributed by **segmenting** the log
/// between `starting`/`stopping mlx runner subprocess` events: every request/timing line belongs to
/// the model whose runner is currently loaded. Exact for serial use (one model at a time); if models
/// run concurrently the untagged lines interleave and per-model metrics become approximate.
public struct OllamaReader: OllamaReading {
    private let logURL: URL
    private let calendar: Calendar

    /// Default macOS location of the Ollama server log.
    public static func defaultLogURL() -> URL {
        URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent(".ollama/logs/server.log")
    }

    /// `calendar` (its time zone) parses the `[GIN]` request timestamps, which carry no zone of
    /// their own. Defaults to the current calendar; tests inject a fixed zone for determinism.
    public init(logURL: URL = OllamaReader.defaultLogURL(), calendar: Calendar = .current) {
        self.logURL = logURL
        self.calendar = calendar
    }

    /// Only the tail of the log is inspected: it can grow to many MB and the live state is at the end.
    private static let tailBytes = 256 * 1024

    public func read() -> OllamaStatus? {
        guard let data = SafeFileReader.tail(at: logURL, maxBytes: Self.tailBytes), !data.isEmpty else {
            return nil
        }
        let text = String(decoding: data, as: UTF8.self)
        let formatter = ginFormatter()

        var segments: [String: Segment] = [:]   // keyed by model tag (dedupes across reloads)
        var order: [String] = []                // first-seen order
        var active: String?                      // model owning the lines being read
        var current: String?                     // model whose runner is still loaded

        for line in text.split(separator: "\n", omittingEmptySubsequences: true) {
            if let tag = startedModel(line) {
                active = tag; current = tag
                if segments[tag] == nil { segments[tag] = Segment(); order.append(tag) }
                continue
            }
            if line.contains("stopping mlx runner subprocess") {
                active = nil; current = nil
                continue
            }
            guard let tag = active, var seg = segments[tag] else { continue }
            apply(line, to: &seg, formatter: formatter)
            segments[tag] = seg
        }

        var status = OllamaStatus()
        status.lastActivity = modificationDate()
        status.models = order.map { tag in
            let s = segments[tag]!
            return OllamaModelActivity(model: tag, current: tag == current, busy: s.busy,
                                       lastActivity: s.lastActivity, contextWindow: s.contextWindow,
                                       contextTokens: s.contextTokens, tokensPerSecond: s.tokensPerSecond,
                                       lastLatency: s.lastLatency, kind: s.kind, requestCount: s.requestCount)
        }
        .sorted { a, b in
            if a.current != b.current { return a.current }   // currently-loaded model first
            return (a.lastActivity ?? .distantPast) > (b.lastActivity ?? .distantPast)
        }
        return status
    }

    // MARK: - Per-model accumulator

    private struct Segment {
        var busy = false
        var lastActivity: Date?
        var contextWindow: Int?
        var contextTokens: Int?
        var tokensPerSecond: Double?
        var lastLatency: TimeInterval?
        var requestCount = 0
        var chatCount = 0
        var embedCount = 0
        var kind: OllamaRequestKind {
            if embedCount > chatCount { return .embed }
            return chatCount > 0 ? .chat : .unknown
        }
    }

    private func apply(_ line: Substring, to seg: inout Segment, formatter: DateFormatter) {
        if let w = intAfter("n_ctx_slot = ", in: line) { seg.contextWindow = w }
        if let t = intAfter("task.n_tokens = ", in: line) { seg.contextTokens = t }
        if let tps = doubleAfter("tg = ", in: line) { seg.tokensPerSecond = tps }
        if line.contains("all slots are idle") { seg.busy = false }
        if line.contains("processing task") { seg.busy = true }
        if line.hasPrefix("[GIN]") { applyGIN(line, to: &seg, formatter: formatter) }
    }

    private func applyGIN(_ line: Substring, to seg: inout Segment, formatter: DateFormatter) {
        guard let endpoint = ginEndpoint(line) else { return }
        let kind: OllamaRequestKind
        if endpoint.contains("/embed") { kind = .embed }
        else if endpoint.hasSuffix("/api/chat") || endpoint.hasSuffix("/api/generate") { kind = .chat }
        else { return }   // /api/tags, /show, /ps, /version, /pull, /delete — not inference
        seg.requestCount += 1
        if kind == .embed { seg.embedCount += 1 } else { seg.chatCount += 1 }
        let fields = line.split(separator: "|", omittingEmptySubsequences: false)
        if fields.count > 2, let dur = parseDuration(fields[2].trimmingCharacters(in: .whitespaces)) {
            seg.lastLatency = dur
        }
        if let first = fields.first {
            let stamp = first.replacingOccurrences(of: "[GIN]", with: "").trimmingCharacters(in: .whitespaces)
            if let date = formatter.date(from: stamp) { seg.lastActivity = date }
        }
    }

    // MARK: - Line parsing helpers

    /// `model=<tag>` on a runner-subprocess *start* line (ignores `pulling manifest model=…` noise).
    private func startedModel(_ line: Substring) -> String? {
        guard line.contains("runner subprocess"), line.contains("starting"),
              let r = line.range(of: "model=") else { return nil }
        let tag = line[r.upperBound...].prefix { !$0.isWhitespace }
        return tag.isEmpty ? nil : String(tag)
    }

    private func intAfter(_ needle: String, in line: Substring) -> Int? {
        guard let r = line.range(of: needle) else { return nil }
        return Int(line[r.upperBound...].prefix { $0.isNumber })
    }

    private func doubleAfter(_ needle: String, in line: Substring) -> Double? {
        guard let r = line.range(of: needle) else { return nil }
        let rest = line[r.upperBound...].drop { $0 == " " }
        return Double(rest.prefix { $0.isNumber || $0 == "." })
    }

    /// The quoted route on a `[GIN]` line, e.g. `/api/chat`.
    private func ginEndpoint(_ line: Substring) -> String? {
        guard let close = line.lastIndex(of: "\"") else { return nil }
        let before = line[..<close]
        guard let open = before.lastIndex(of: "\"") else { return nil }
        return String(line[line.index(after: open)..<close])
    }

    /// llama.cpp's Go-duration strings: `5.0s`, `200ms`, `23.917µs`, `41ns`.
    private func parseDuration(_ s: String) -> TimeInterval? {
        func value(_ drop: Int, _ scale: Double) -> TimeInterval? { Double(s.dropLast(drop)).map { $0 * scale } }
        if s.hasSuffix("ms") { return value(2, 1e-3) }
        if s.hasSuffix("µs") || s.hasSuffix("us") { return value(2, 1e-6) }
        if s.hasSuffix("ns") { return value(2, 1e-9) }
        if s.hasSuffix("s") { return value(1, 1) }
        return nil
    }

    private func ginFormatter() -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.calendar = Calendar(identifier: .gregorian)
        f.timeZone = calendar.timeZone
        f.dateFormat = "yyyy/MM/dd - HH:mm:ss"
        return f
    }

    private func modificationDate() -> Date? {
        (try? logURL.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate
    }
}
